---
title: "Undercaffeinated_And_Underpaid Final Project - R Notebook"
output: html_notebook
---
Install required packages.
```{r}
install.packages('Rcpp', dependencies = TRUE)
install.packages('gplots')
install.packages('Biobase')
install.packages('multtest')
install.packages('LDheatmap')
install.packages('genetics')
install.packages('compiler')
install.packages('scatterplot3d')
```
Load required packages.
```{r}
library(Rcpp)
library(gplots)
library(Biobase)
library(multtest)
library(LDheatmap)
library(genetics)
library(compiler)
library(scatterplot3d)
source("https://bioconductor.org/biocLite.R")
biocLite("Biobase", "multtest")
```
Load functions from GAPIT.
```{r}
source("http://www.zzlab.net/GAPIT/gapit_functions.txt")
source("http://www.zzlab.net/GAPIT/emma.txt")
```
Loads genotypic data file.
```{r}
GD <- read.table("all_GD.txt", header = T) 
```
Loads genetic map.
```{r}
myGM <- read.table("all_GM.txt",header = T)
```
Converts the chromosome number and removes "GM" from the names.
```{r}
myGM$chr <- as.numeric(substr(myGM$chr,3,4))
```
Load names of phenotypes and makes a directory for GWAS output.
```{r}
myYs <- list.files("E:/BCB546X-Fall2017_Undercaffeinated_And_Underpaid/Data/Phenotype_Data") 
results_dir <- "E:/BCB546X-Fall2017_Undercaffeinated_And_Underpaid/Results"
```
Run GWAS on all traits with for loop.
```{r}
for (i in 1:length(myYs)) {
  setwd("E:/BCB546X-Fall2017_Undercaffeinated_And_Underpaid/Data/Phenotype_Data/") # set directory where the phenotype files are located
  Y <- read.csv(file = myYs[i]) # read in next set of trait phenotypes
  y <- colnames(Y)[2] # obtain the name of the trait. All files have phenotypes in second column (starting from 1, not 0)
  line = Y[names(Y) %in% c('acid', y)] # obtain list of acession identifiers found in the phenotype file
  line1 <- line[(line$acid %in% GD$PI),] # filter the phenotype file for only accessions that have been genotyped (those in genotype file)
  myY <- line1[which(line1$acid %in% GD$PI),] # select lines that have been genotyped
  myGD <- GD[which(GD$PI %in% myY$acid),]# select lines that have observations available
  g= nrow(myY) # number of accessions in current GWAS test... to be used as an argument in the "GAPIT" function
  dir.create(paste(results_dir, y)) # create a folder withing the "Results" directory to store the output of the GWAS test
  setwd(paste(results_dir, y)) # change the directory to output the GAPIT results into the folder just created
  myGAPIT <- GAPIT(Y = myY, 
                   GD = myGD, 
                   GM = myGM, 
                   PCA.total = 3, 
                   group.from = g, 
                   group.to = g,
                   SNP.fraction = 1, 
                   SNP.MAF = 0.05) # regular MLM with no compression
}
```
Complies all of the significant SNPs for all traits.
```{r}
setwd("/Users/jmshook/Desktop/Beagle GWAS/Dx")
files <- list.files(pattern = "\\.GWAS.Results.csv$", recursive = TRUE)
data2=lapply(files, read.table, header=T, sep=",")
for (i in 1:length(data2)){data2[[i]]<-cbind(data2[[i]],files[i])}
data_rbind <- do.call("rbind", data2) 
colnames(data_rbind)[c(1:10)]<-c("SNP", "Chromosome", "Position", "P.value", "maf", "nobs", "Rsquare.of.Model.without.SNP", "Rsquare.of.Model.with.SNP", "`FDR_Adjusted_P-values`","file.name")
signif <- data_rbind[(data_rbind[,9]<0.05),]
write.csv(signif,'SignificantSNPsForDiseases.csv')
```




