---
title: "Undercaffeinated_And_Underpaid Final Project - R Notebook"
output: html_notebook
---
This code documents the whole R pipeline for running the GWAS. 

The code below was developed to transfer the format and imputate the missing datapoints 
of the SNP data downloaded from the Soybase to form GD and GM files that work for 
GAPIT. 

Install needed packages.
```{r}
install.packages("synbreed")
install.packages("gdata")
```

Read in VCF from Soybase, shuffle, label, and examines the file.
```{r}
setwd("E:/BCB546X-Fall2017_Undercaffeinated_And_Underpaid")
GD <- read.vcf2matrix("fullpanel.vcf.gz", cores=20)
dim(GD)
head(GD)[,1:5]
tail(GD)[,1:5]
GD <- GD[c(3,1,2,4:ncol(GD))]
names(GD)[1:3] <- c("SNP","Chromosome","Position")
dim(GD)
```
Remove markers on scaffolds, shift columns, and converts chromosome keys to numbers.
```{r}
GD <- GD[which(nchar(as.character(GD$Chrom)) < 5),]
dim(GD)
table(GD$Chrom)
GD$chrom <- as.numeric(substr(GD$Chrom,3,4))
GD <- GD[c(1,ncol(GD),3:(ncol(GD)-1))]
GD <- GD[order(GD$chrom,GD$Position),]
head(GD)[,1:5]
tail(GD)[,1:5]
```
Convert letter-based SNPs to numeric for analysis.
```{r}
convert.snp <- function(x) 
  {
  alleles <- sort(unique(x))
  y <- rep(NA, length(x))
  y[which(x == "A")] <- "A/A"
  y[which(x == "T")] <- "T/T"
  y[which(x == "C")] <- "C/C"
  y[which(x == "G")] <- "G/G"
  
  if ("H" %in% alleles)
{
  if (which(alleles=="H")==3) 
  {
    y[which(x == "H")] <- paste(alleles[1], alleles[2], sep = "/")
  }
  else
  {
    if ((which(alleles=="H")==2) & ("T" %in% alleles))
    {
      y[which(x == "H")] <- paste(alleles[1], alleles[3], sep = "/")
    }
    else
    {
      if ((which(alleles=="H")==2) & !("T" %in% alleles))
      {y[which(x == "H")] <- paste(alleles[1], "T", sep = "/")}
      else {y[which(x == "H")] <- "A/T"}  
     }
   }
 }
  return(y)
}
```
Examine format of genotype file.
```{r}
dim(geno)
class(geno)
```
Label genotype rows and columns
```{r}
rownames(geno) <- colnames(GD[4:ncol(GD)])
colnames(geno) <- GD[1:nrow(GD),1]
```
View dataframes
```{r}
GD[1:5,1:8]
geno[1:5,1:5]
```
Create Mapfile of SNPs
```{r}
Map <- GD[2:3]
class(Map)
colnames(Map) <- c("chr", "pos")
Map[1:4,1:2]
rownames(Map) <- colnames(geno)
```
Perform Beagle imputation from the comfort of R
```{r}
setwd("E:/BCB546X-Fall2017_Undercaffeinated_And_Underpaid")
gData<- create.gpData(pheno=NULL, geno=geno, map=Map, covar= NULL, reorderMap=F,map.unit="bp")
gData_coded <- codeGeno(gData, maf= 0.01, nmiss= 0.1, impute=T, impute.type="beagle", label.heter="alleleCoding")
```
Examine data and subset out the needed information to build a map
```{r}
dim(gData$map)
dim(gData_coded$map)
gData_coded$map[1:5,1:2]
gData_coded$map$SNP <- colnames(gData_coded$geno)
dim(gData_coded$map)
filted_Map <- gData_coded$map[c(3,1,2)]
filted_Map[1:5,1:3]
```
Create map file for use in GWAS
```{r}
setwd("E:/BCB546X-Fall2017_Undercaffeinated_And_Underpaid")
write.table(filted_Map, "all_GM.txt", row.names=F,quote=F, sep="\t")
```
Examine data and subset out the needed information to compile all markers
```{r}
dim(gData$geno)
dim(gData_coded$geno)
gData$geno[180:nrow(gData$geno),1:5]
gData_coded$geno[180:nrow(gData$geno),1:5]
str(gData_coded$geno)
class(gData_coded$geno)
```
Create Marker file
```{r}
geno_df <- as.data.frame(gData_coded$geno)
geno_df$PI <- rownames(geno_df)
cn <- ncol(geno_df)
geno_df[1:5,(cn-5):cn]
geno_df <- geno_df[c(cn,1:(cn-1))]
geno_df[1:5,1:5]
write.table(geno_df,row.names=F, quote=F, "all_GD.txt",sep="\t")
```
GWAS Code

Install required packages.
```{r}
install.packages('Rcpp', dependencies = TRUE)
install.packages('gplots')
install.packages('Biobase')
install.packages('multtest')
install.packages('LDheatmap')
install.packages('genetics')
install.packages('compiler')
install.packages('scatterplot3d')
```
Load required packages.
```{r}
library(Rcpp)
library(gplots)
library(Biobase)
library(multtest)
library(LDheatmap)
library(genetics)
library(compiler)
library(scatterplot3d)
source("https://bioconductor.org/biocLite.R")
biocLite("Biobase", "multtest")
```
Load functions from GAPIT.
```{r}
source("http://www.zzlab.net/GAPIT/gapit_functions.txt")
source("http://www.zzlab.net/GAPIT/emma.txt")
```
Loads genotypic data file.
```{r}
GD <- read.table("all_GD.txt", header = T) 
```
Loads genetic map.
```{r}
myGM <- read.table("all_GM.txt",header = T)
```
Converts the chromosome number and removes "GM" from the names.
```{r}
myGM$chr <- as.numeric(substr(myGM$chr,3,4))
```
Load names of phenotypes and makes a directory for GWAS output.
```{r}
myYs <- list.files("E:/BCB546X-Fall2017_Undercaffeinated_And_Underpaid/Phenotype_Files/") 
results_dir <- "E:/BCB546X-Fall2017_Undercaffeinated_And_Underpaid/Results/"
```
Run GWAS on all traits with for loop.
```{r}
for (i in 1:length(myYs)) {
  setwd("E:/BCB546X-Fall2017_Undercaffeinated_And_Underpaid/Phenotype_Files/") # set directory where the phenotype files are located
  Y <- read.csv(file = myYs[i]) # read in next set of trait phenotypes
  y <- colnames(Y)[2] # obtain the name of the trait. All files have phenotypes in second column (starting from 1, not 0)
  line = Y[names(Y) %in% c('acid', y)] # obtain list of acession identifiers found in the phenotype file
  line1 <- line[(line$acid %in% GD$PI),] # filter the phenotype file for only accessions that have been genotyped (those in genotype file)
  myY <- line1[which(line1$acid %in% GD$PI),] # select lines that have been genotyped
  myGD <- GD[which(GD$PI %in% myY$acid),]# select lines that have observations available
  g= nrow(myY) # number of accessions in current GWAS test... to be used as an argument in the "GAPIT" function
  dir.create(paste(results_dir, y)) # create a folder withing the "Results" directory to store the output of the GWAS test
  setwd(paste(results_dir, y)) # change the directory to output the GAPIT results into the folder just created
  myGAPIT <- GAPIT(Y = myY, 
                   GD = myGD, 
                   GM = myGM, 
                   PCA.total = 3, 
                   group.from = g, 
                   group.to = g,
                   SNP.fraction = 1, 
                   SNP.MAF = 0.05) # regular MLM with no compression
}
```
Complies all of the significant SNPs for all traits.  Runs on personal computer, can modify working directory to run elsewhere.
```{r}
setwd("E:/BCB546X-Fall2017_Undercaffeinated_And_Underpaid/Dx")
files <- list.files(pattern = "\\.GWAS.Results.csv$", recursive = TRUE)
data2=lapply(files, read.table, header=T, sep=",")
for (i in 1:length(data2)){data2[[i]]<-cbind(data2[[i]],files[i])}
data_rbind <- do.call("rbind", data2) 
colnames(data_rbind)[c(1:10)]<-c("SNP", "Chromosome", "Position", "P.value", "maf", "nobs", "Rsquare.of.Model.without.SNP", "Rsquare.of.Model.with.SNP", "`FDR_Adjusted_P-values`","file.name")
signif <- data_rbind[(data_rbind[,9]<0.05),]
write.csv(signif,'SignificantSNPsForDiseases.csv')
```




